kod/util/gameevent/invasion/jonasatt.kod

```diff
@@ -0,0 +1,298 @@
+      {
+         plSpawnRects = [ [[27,10],[13,5]], [[36,7],[42,16]],
+                          [[45,30],[37,25]], [[45,41],[23,22]] ];
+      }
+
+      propagate;
+   }
+
+   Update()
+   {
+      if piPhase = 1
+      {
+         if Send(self,@CountMonsters,#type=&RebelTroop) = 0
+         {
+            Send(self,@PhaseTwo);
+         }
+      }
+
+      if piPhase = 2
+      {
+         if Send(self,@CountMonsters,#type=&Troll) = 0
+            AND Send(self,@CountMonsters,#type=&RebelTroop) = 0
+         {
+            Send(self,@PhaseThree);
+         }
+      }
+
+      if piPhase = 3
+      {
+         if Send(self,@CountMonsters,#type=&PrincessTroop) = 0
+            AND Send(self,@CountMonsters,#type=&Troll) = 0
+            AND Send(self,@CountMonsters,#type=&RebelTroop) = 0
+         {
+            Send(self,@FinalPhase);
+         }
+      }
+
+      if piPhase = 4
+      {
+         if Send(self,@CountMonsters,#type=&DarkAngel) = 0
+         {
+            % Call EventEnd to handle ending the event properly. This
+            % deletes the event in the GameEventEngine, which also calls
+            % the appropriate code in EndEvent, which handles the end of
+            % the specific event.
+            Send(self,@EventEnd);
+         }
+      }
+
+      propagate;
+   }
+
+   StartEvent()
+   {
+      Send(self,@PhaseOne);
+
+      propagate;
+   }
+
+   PhaseOne()
+   {
+      local i,lSpawnPoint;
+
+      piPhase = 1;
+
+      i = piRebelTroops;
+
+      while (i > 0)
+      {
+         lSpawnPoint = Send(self,@MakeSpawnPoint);
+         Send(Send(SYS,@FindRoomByNum,#num=piActiveRoom),
+               @NewHold,#what=Create(&RebelTroop),
+               #new_row=Nth(lSpawnPoint,1),#new_col=Nth(lSpawnPoint,2));
+         i = i - 1;
+      }
+
+      return;
+   }
+
+   PhaseTwo()
+   {
+      local i,lSpawnPoint;
+
+      piPhase = 2;
+
+      if vbAnnounce
+      {
+         Send(SYS,@SystemBroadcast,#type=SAY_MESSAGE,
+               #string=JonasAttack_phase2_msg);
+      }
+
+      i = piRebelTroops;
+
+      while (i > 0)
+      {
+         lSpawnPoint = Send(self,@MakeSpawnPoint);
+         Send(Send(SYS,@FindRoomByNum,#num=piActiveRoom),
+               @NewHold,#what=Create(&RebelTroop),
+               #new_row=Nth(lSpawnPoint,1),#new_col=Nth(lSpawnPoint,2));
+         i = i - 1;
+      }
+
+      i = piTrolls;
+
+      while (i > 0)
+      {
+         lSpawnPoint = Send(self,@MakeSpawnPoint);
+         Send(Send(SYS,@FindRoomByNum,#num=piActiveRoom),
+               @NewHold,#what=Create(&Troll),
+               #new_row=Nth(lSpawnPoint,1),#new_col=Nth(lSpawnPoint,2));
+         i = i - 1;
+      }
+
+      return;
+   }
+
+   PhaseThree()
+   {
+      local i,lSpawnPoint;
+
+      piPhase = 3;
+
+      if vbAnnounce
+      {
+         Send(SYS,@SystemBroadcast,#type=SAY_MESSAGE,
+               #string=JonasAttack_phase3_msg);
+      }
+
+      i = piRebelTroops;
+
+      while (i > 0)
+      {
```

Comment:
This is where you will call your new ChangeRoom message, just before you pick the spawn point and make a monster.